crates/runix/src/installable.rs

```diff
@@ -1,45 +1,262 @@
 //! A much simplified installable representation
 
-use derive_more::Constructor;
+use std::fmt::Display;
+use std::str::FromStr;
 
-/// A much simplified installable representation
-#[derive(Debug, Clone, Constructor)]
+use once_cell::sync::Lazy;
+use regex::Regex;
+use thiserror::Error;
+
+use crate::flake_ref::{FlakeRef, ParseFlakeRefError};
+
+/// regex describing valid characters for attributes
+/// derived from <https://github.com/flox/nix/blob/ysndr/disable_attrpath_resolution/src/libutil/url-parts.hh#L21>
+static VALID_ATTRIBUTE: Lazy<Regex> =
+    Lazy::new(|| Regex::new("^([a-zA-Z0-9-._~!$&'()*+,;=:@?/ ]*)$").unwrap());
+
+/// A simplified installable representation
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub struct Installable {
     pub flakeref: FlakeRef,
-    pub attr_path: String,
+    pub attr_path: AttrPath,
 }
 
-impl Installable {
-    pub fn to_nix(&self) -> String {
-        format!("{}#{}", self.flakeref, self.attr_path)
+/// The attrpath component of an installable
+///
+/// This implementation wraps a [Vec<String>] for components.
+/// Its [FromStr] and [Display] implementations are used to parse and validate,
+/// as well as canonically print an attrpath.
+///
+/// The implementation also includes [TryFrom] implementations to allow
+/// ergonomic validated conversion from list types
+///
+/// ```
+/// use runix::installable::AttrPath;
+///
+/// "abc.xyz".parse::<AttrPath>().expect("Parses from String");
+/// AttrPath::try_from(["abc", "xyz"]).expect("Parses from array");
+/// AttrPath::try_from(&*vec!["abc", "xyz"]).expect("Parses from vec");
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub struct AttrPath(Vec<Attribute>);
+
+impl AttrPath {
+    /// Add another component to the end of the attrpath
+    pub fn push_attr(&mut self, attr: &str) -> Result<&mut Self, <Attribute as FromStr>::Err> {
+        self.0.push(attr.parse::<Attribute>()?);
+        Ok(self)
+    }
+
+    /// get an iterator over all components of the attrpath
+    pub fn components(&self) -> impl Iterator<Item = &Attribute> {
+        self.0.iter()
+    }
+
+    /// determine whether the attrpath is empty
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
     }
 }
 
-impl From<String> for Installable {
-    fn from(input: String) -> Self {
-        let mut split = input.splitn(2, '#');
+impl FromStr for AttrPath {
+    type Err = ParseInstallableError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut attributes = AttrPath::default();
+        let mut cur = String::new();
+
+        let mut start_quote = None;
+
+        for (n, c) in s.chars().enumerate() {
+            match c {
+                '.' if start_quote.is_none() => {
```

Comment:
Are you deliberately allowing supporting quoting part of an attribute? E.g. `a"b"c` becomes `abc`. Maybe add that as a test case

Reply:
Nix seems to yeah
I'll add a comment linking to thatbpart in the nix source

Reply:
Nix seems to yeah
I'll add a comment linking to the counter part in the nix source .